h/atom_feed.py

```diff
@@ -0,0 +1,173 @@
+        "{text}".format(
+            selection=cgi.escape(get_selection(annotation)),
+            text=cgi.escape(annotation["text"])))
+
+    entry["links"] = []
+
+    entry["links"].append({"rel": "alternate", "type": "text/html",
+                           "href": annotation_url(annotation)})
+
+    if annotation_api_url:
+        entry["links"].append({"rel": "alternate", "type": "application/json",
+                               "href": annotation_api_url(annotation)})
+
+    return entry
+
+
+def _feed_from_annotations(
+        annotations, atom_url, annotation_url, annotation_api_url=None,
+        html_url=None, title=None, subtitle=None):
+    """Return an Atom feed for the given list of annotations.
+
+    This returns a logical representation of an Atom feed as a Python dict
+    containing all of the data that a template would need to render the feed
+    to XML (including a list of dicts for the feed's entries).
+
+    """
+    links = [{"rel": "self", "type": "application/atom+xml", "href": atom_url}]
+
+    if html_url:
+        links.append(
+            {"rel": "alternate", "type": "text/html", "href": html_url})
+
+    entries = [
+        _feed_entry_from_annotation(a, annotation_url, annotation_api_url)
+        for a in annotations]
+
+    feed = {
+        "id": atom_url,
+        "title": title or _("Hypothesis Stream"),
+        "subtitle": subtitle or _("The Web. Annotated"),
+        "entries": entries,
+        "links": links
+    }
+
+    if annotations:
+        feed["updated"] = annotations[0]["updated"]
+
+    return feed
+
+
+def render(
+        request, annotations, atom_url, html_url, title=None, subtitle=None):
+    """Return an Atom feed of the given list of annotations.
+
+    :param annotations: A list of annotations from the API
+    :type annotations: list of dicts
+
+    :param atom_url: The URL where this Atom feed will be hosted
+        (the feed will contain a link to this URL)
+    :type atom_url: unicode
+
+    :param html_url: The URL to the HTML page that this Atom feed is a feed
+        for (the feed will contain a link to this URL).
+    :type html_url: unicode
+
+    :param title: The title of the feed
+    :type title: unicode
+
+    :param subtitle: The subtitle of the feed
+    :type subtitle: unicode
+
+    :returns: An Atom feed as an XML string
+    :rtype: unicode
+
+    """
+    def annotation_url(annotation):
+        """Return the HTML permalink URL for the given annotation."""
+        return request.resource_url(request.root, "a", annotation["id"])
+
+    def annotation_api_url(annotation):
+        """Return the JSON API URL for the given annotation."""
+        return request.resource_url(request.root, "api", "annotation",
+                                    annotation["id"])
+
+    feed = _feed_from_annotations(
+        annotations=annotations, atom_url=atom_url,
+        annotation_url=annotation_url, annotation_api_url=annotation_api_url,
```

Comment:
The closure is a nice way out.

We'd have to add the routes. I'm convinced it's a good idea.
On Mar 20, 2015 3:06 PM, "Sean Hammond" <notifications@github.com> wrote:

> In h/atom_feed.py
> <https://github.com/hypothesis/h/pull/2029#discussion_r26867474>:
>
> > +    :returns: An Atom feed as an XML string
> > +    :rtype: unicode
> > +
> > +    """
> > +    def annotation_url(annotation):
> > +        """Return the HTML permalink URL for the given annotation."""
> > +        return request.resource_url(request.root, "a", annotation["id"])
> > +
> > +    def annotation_api_url(annotation):
> > +        """Return the JSON API URL for the given annotation."""
> > +        return request.resource_url(request.root, "api", "annotation",
> > +                                    annotation["id"])
> > +
> > +    feed = _feed_from_annotations(
> > +        annotations=annotations, atom_url=atom_url,
> > +        annotation_url=annotation_url, annotation_api_url=annotation_api_url,
>
> Do we have those named routes already, or would we have to add them?
>
> request.resource_url(request.root, ...) is a pain. It's difficult to mock
> in tests. Even a DummyRequest or even an actual Request object does not
> have a .root property unless there's an actual request (i.e. you're using
> webtest and simulating an HTTP request). And .root is not exactly nice to
> mock afaict, having to be a hierarchy of resource objects.
>
> request.route_url('annotation', id='...') seems a lot easier to mock.
>
> But I agree let's leave this for now. This problem was what drove me
> eventually to the closure-based solution above which made me feel very
> clever :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/hypothesis/h/pull/2029/files#r26867474>.
>